File: ppai/audio.py
"""击球声检测。

思路：乒乓球击球是一个极短的宽带瞬态（~5-15ms），能量集中在 1-5kHz，
起振非常陡。人声、球馆嗡鸣、空调这类干扰要么频率低、要么起振缓慢，
所以在限定频带上做「谱通量（spectral flux）」+ 自适应阈值就能把它挑出来，
不需要神经网络。这是第一版；升级路线是 YAMNet / PANNs。
"""
from __future__ import annotations

import subprocess
from typing import Dict, Tuple

import numpy as np


def extract_pcm(video_path: str, sr: int) -> np.ndarray:
    """用 ffmpeg 抽单声道 float32 PCM。无音轨时返回空数组。"""
    cmd = [
        "ffmpeg", "-v", "error", "-i", video_path,
        "-vn", "-ac", "1", "-ar", str(sr), "-f", "f32le", "-",
    ]
    proc = subprocess.run(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    if proc.returncode != 0 or not proc.stdout:
        return np.zeros(0, dtype=np.float32)
    return np.frombuffer(proc.stdout, dtype=np.float32)


def onset_envelope(x: np.ndarray, cfg: Dict) -> Tuple[np.ndarray, float]:
    """返回 (谱通量包络, 帧率)。分块处理，一小时视频也只占几 MB 内存。"""
    sr, n_fft, hop = cfg["sr"], cfg["n_fft"], cfg["hop"]
    win = np.hanning(n_fft).astype(np.float32)
    freqs = np.fft.rfftfreq(n_fft, 1.0 / sr)
    band = (freqs >= cfg["fmin"]) & (freqs <= cfg["fmax"])

    frame_rate = sr / float(hop)
    if len(x) < n_fft:
        return np.zeros(0, dtype=np.float32), frame_rate
    n_frames = 1 + (len(x) - n_fft) // hop

    env = np.zeros(n_frames, dtype=np.float32)
    offsets = np.arange(n_fft)
    prev_tail = None
    block = 4000

    for s in range(0, n_frames, block):
        e = min(s + block, n_frames)
        idx = (np.arange(s, e) * hop)[:, None] + offsets[None, :]
        frames = x[idx] * win
        mag = np.abs(np.fft.rfft(frames, axis=1)).astype(np.float32)[:, band]
        # 对数压缩：让检测对整体音量不敏感，只看相对突变
        logmag = np.log1p(mag * 100.0)
        head = prev_tail if prev_tail is not None else logmag[0]
        shifted = np.vstack([head[None, :], logmag[:-1]])
        # 只取正差分 —— 我们要的是「突然出现的能量」，不是衰减
        env[s:e] = np.maximum(logmag - shifted, 0.0).sum(axis=1)
        prev_tail = logmag[-1]

    return env, frame_rate


def _adaptive_threshold(env: np.ndarray, frame_rate: float, win_s: float, k: float) -> np.ndarray:
    """局部中位数 + k*MAD。球馆底噪基本平稳，用鲁棒统计比全局阈值稳得多。

    按整块算统计量再线性插值，避免 O(N*W) 的滑动中位数。
    """
    w = max(1, int(round(win_s * frame_rate)))
    n_blocks = max(1, int(np.ceil(len(env) / w)))
    pad = n_blocks * w - len(env)
    padded = np.concatenate([env, np.repeat(env[-1:], pad)]) if pad else env
    blocks = padded.reshape(n_blocks, w)

    med = np.median(blocks, axis=1)
    mad = np.median(np.abs(blocks - med[:, None]), axis=1) * 1.4826

    centers = np.arange(n_blocks) * w + w / 2.0
    grid = np.arange(len(env), dtype=np.float64)
    med_i = np.interp(grid, centers, med)
    mad_i = np.interp(grid, centers, mad)
    # MAD 可能为 0（极安静段），给个下限防止阈值塌到 0 触发满屏误检
    floor = np.percentile(env, 60) * 0.05 if len(env) else 0.0
    return med_i + k * np.maximum(mad_i, floor)


def detect_hits(x: np.ndarray, cfg: Dict) -> Tuple[np.ndarray, np.ndarray, np.ndarray, float]:
    """返回 (击球时间点, 包络, 阈值曲线, 帧率)。"""
    env, frame_rate = onset_envelope(x, cfg)
    if len(env) == 0:
        return np.zeros(0), env, np.zeros(0), frame_rate

    thr = _adaptive_threshold(env, frame_rate, cfg["noise_win_s"], cfg["k_mad"])

    # 局部极大值且超过阈值
    is_peak = np.zeros(len(env), dtype=bool)
    is_peak[1:-1] = (env[1:-1] >= env[:-2]) & (env[1:-1] > env[2:]) & (env[1:-1] > thr[1:-1])
    cand = np.flatnonzero(is_peak)
    if len(cand) == 0:
        return np.zeros(0), env, thr, frame_rate

    # 按强度降序贪心去重，保证最小间隔
    min_gap = max(1, int(round(cfg["min_gap_s"] * frame_rate)))
    order = cand[np.argsort(-env[cand])]
    kept = []
    taken = np.zeros(len(env), dtype=bool)
    for i in order:
        lo, hi = max(0, i - min_gap), min(len(env), i + min_gap + 1)
        if not taken[lo:hi].any():
            kept.append(i)
            taken[i] = True

    hits = np.sort(np.array(kept)) / frame_rate
    return hits, env, thr, frame_rate

File: android/core/src/main/kotlin/cc/pipo/core/Audio.kt
package cc.pipo.core

import kotlin.math.abs
import kotlin.math.cos
import kotlin.math.ln1p
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt
import kotlin.math.sqrt

/**
 * 击球声检测 —— ppai/audio.py 的 Kotlin 移植。
 *
 * 这是整个产品的地基：回合分组、主题排序、成片时长全都建在它输出的
 * 那串时间点上。所以它**必须和 Python 逐点一致**，不是「差不多就行」。
 * 同目录的 AudioParityTest 拿 Python 在真实素材上的输出做基准逐点比对。
 *
 * 移植时几处必须和 numpy 对齐的细节（都踩过或差点踩到）：
 *
 *  * `np.hanning(N)` 是**对称**窗 `0.5 - 0.5*cos(2πn/(N-1))`，
 *    不是信号处理里常见的周期窗 `2πn/N`。差一个样本，谱就对不上。
 *  * `np.median` 在偶数长度上取中间两个的**平均**，不是下中位数。
 *  * `np.percentile(x, 60)` 用线性插值（numpy 默认），不是取最近秩。
 *  * `np.interp` 在两端**夹住**（用首/末值），不外推。
 *  * 峰值判据左闭右开：`env[i] >= env[i-1] && env[i] > env[i+1]`，
 *    两边都用 `>` 会漏掉平顶峰，都用 `>=` 会把平台段整段选中。
 *
 * 不依赖 Android，纯 JVM —— 这样单测不用模拟器就能跑，改一行五秒钟知道对错。
 */
object Audio {

    data class Config(
        val sr: Int = 16000,
        val nFft: Int = 512,
        val hop: Int = 160,
        val fmin: Double = 800.0,
        val fmax: Double = 6000.0,
        val kMad: Double = 4.0,
        val noiseWinS: Double = 2.0,
        val minGapS: Double = 0.06,
    )

    data class Result(
        val hits: DoubleArray,      // 击球时刻（秒）
        val env: FloatArray,        // 谱通量包络
        val thr: FloatArray,        // 自适应阈值曲线
        val frameRate: Double,
    )

    /** 谱通量包络。分块处理，一小时素材也只占几 MB。 */
    fun onsetEnvelope(x: FloatArray, c: Config): Pair<FloatArray, Double> {
        val frameRate = c.sr.toDouble() / c.hop
        if (x.size < c.nFft) return FloatArray(0) to frameRate

        // 对称 Hann，和 np.hanning 一致
        val win = FloatArray(c.nFft) {
            (0.5 - 0.5 * cos(2.0 * Math.PI * it / (c.nFft - 1))).toFloat()
        }
        // rfftfreq(nFft, 1/sr)[k] = k * sr / nFft
        val nBins = c.nFft / 2 + 1
        var lo = -1
        var hi = -1
        for (k in 0 until nBins) {
            val f = k.toDouble() * c.sr / c.nFft
            if (f >= c.fmin && f <= c.fmax) {
                if (lo < 0) lo = k
                hi = k
            }
        }
        if (lo < 0) return FloatArray(0) to frameRate
        val bandLen = hi - lo + 1

        val nFrames = 1 + (x.size - c.nFft) / c.hop
        val env = FloatArray(nFrames)
        val fft = Fft(c.nFft)
        val re = DoubleArray(c.nFft)
        val im = DoubleArray(c.nFft)
        var prev: DoubleArray? = null
        val cur = DoubleArray(bandLen)

        for (t in 0 until nFrames) {
            val off = t * c.hop
            for (i in 0 until c.nFft) {
                re[i] = (x[off + i] * win[i]).toDouble()
                im[i] = 0.0
            }
            fft.transform(re, im)
            for (k in 0 until bandLen) {
                val b = lo + k
                val mag = sqrt(re[b] * re[b] + im[b] * im[b])
                cur[k] = ln1p(mag * 100.0)   // 对数压缩：只看相对突变，不看绝对音量
            }
            // 第一帧没有前一帧，Python 用它自己当 head，所以通量恒为 0
            val p = prev
            var sum = 0.0
            if (p != null) {
                for (k in 0 until bandLen) {
                    val d = cur[k] - p[k]
                    if (d > 0) sum += d          // 只要「突然出现的能量」，不要衰减
                }
            }
            env[t] = sum.toFloat()
            prev = cur.copyOf()
        }
        return env to frameRate
    }

    /** 局部中位数 + k·MAD。按块算再线性插值，避免 O(N·W) 的滑动中位数。 */
    fun adaptiveThreshold(env: FloatArray, frameRate: Double, winS: Double, k: Double): FloatArray {
        val n = env.size
        if (n == 0) return FloatArray(0)
        val w = max(1, (winS * frameRate).roundToInt())
        val nBlocks = max(1, Math.ceil(n.toDouble() / w).toInt())
        val padded = DoubleArray(nBlocks * w) { if (it < n) env[it].toDouble() else env[n - 1].toDouble() }

        val med = DoubleArray(nBlocks)
        val mad = DoubleArray(nBlocks)
        val buf = DoubleArray(w)
        for (b in 0 until nBlocks) {
            System.arraycopy(padded, b * w, buf, 0, w)
            val m = median(buf.copyOf())
            med[b] = m
            for (i in 0 until w) buf[i] = abs(padded[b * w + i] - m)
            mad[b] = median(buf.copyOf()) * 1.4826
        }

        val centers = DoubleArray(nBlocks) { it * w + w / 2.0 }
        // MAD 可能为 0（极安静段），给下限，否则阈值塌到 0 会满屏误检
        val floor = percentile(env, 60.0) * 0.05
        return FloatArray(n) { i ->
            val m = interp(i.toDouble(), centers, med)
            val a = interp(i.toDouble(), centers, mad)
            (m + k * max(a, floor)).toFloat()
        }
    }

    fun detectHits(x: FloatArray, c: Config): Result {
        val (env, frameRate) = onsetEnvelope(x, c)
        if (env.isEmpty()) return Result(DoubleArray(0), env, FloatArray(0), frameRate)
        val thr = adaptiveThreshold(env, frameRate, c.noiseWinS, c.kMad)

        // 局部极大值且超过阈值
        val cand = ArrayList<Int>()
        for (i in 1 until env.size - 1) {
            if (env[i] >= env[i - 1] && env[i] > env[i + 1] && env[i] > thr[i]) cand.add(i)
        }
        if (cand.isEmpty()) return Result(DoubleArray(0), env, thr, frameRate)

        // 按强度降序贪心去重，保证最小间隔。
        // 并列时按帧号升序 —— numpy 的 argsort 默认是 quicksort（不稳定），
        // 真出现完全相同的通量值时两边可能选不同的那个。实测基准里没有并列，
        // 这里定死顺序是为了让 Kotlin 侧自己可复现。
        val minGap = max(1, (c.minGapS * frameRate).roundToInt())
        val order = cand.sortedWith(compareByDescending<Int> { env[it] }.thenBy { it })
        val taken = BooleanArray(env.size)
        val kept = ArrayList<Int>()
        for (i in order) {
            val a = max(0, i - minGap)
            val b = min(env.size, i + minGap + 1)
            var clash = false
            for (j in a until b) if (taken[j]) { clash = true; break }
            if (!clash) { kept.add(i); taken[i] = true }
        }
        kept.sort()
        return Result(DoubleArray(kept.size) { kept[it] / frameRate }, env, thr, frameRate)
    }

    /** 每次击球的通量峰值 —— 近似「力量」。取峰值附近最大值，容忍几帧定位误差。 */
    fun hitAmplitudes(hits: DoubleArray, env: FloatArray, frameRate: Double): DoubleArray {
        if (env.isEmpty() || hits.isEmpty()) return DoubleArray(hits.size)
        return DoubleArray(hits.size) { n ->
            val i = (hits[n] * frameRate).toInt().coerceIn(0, env.size - 1)
            var m = Double.NEGATIVE_INFINITY
            for (j in max(0, i - 2)..min(env.size - 1, i + 2)) m = max(m, env[j].toDouble())
            m
        }
    }

    // ── numpy 语义的小工具 ─────────────────────────────────

    /** np.median：偶数长度取中间两个的平均。会就地排序传入的数组。 */
    internal fun median(a: DoubleArray): Double {
        if (a.isEmpty()) return 0.0
        a.sort()
        val n = a.size
        return if (n % 2 == 1) a[n / 2] else (a[n / 2 - 1] + a[n / 2]) / 2.0
    }

    /** np.percentile 的默认线性插值法。 */
    internal fun percentile(x: FloatArray, q: Double): Double {
        if (x.isEmpty()) return 0.0
        val a = DoubleArray(x.size) { x[it].toDouble() }
        a.sort()
        val pos = (q / 100.0) * (a.size - 1)
        val lo = Math.floor(pos).toInt()
        val hi = Math.ceil(pos).toInt()
        if (lo == hi) return a[lo]
        return a[lo] + (a[hi] - a[lo]) * (pos - lo)
    }

    /** np.interp：xp 递增，两端夹住不外推。 */
    internal fun interp(v: Double, xp: DoubleArray, fp: DoubleArray): Double {
        if (xp.isEmpty()) return 0.0
        if (xp.size == 1 || v <= xp[0]) return fp[0]
        if (v >= xp[xp.size - 1]) return fp[fp.size - 1]
        var lo = 0
        var hi = xp.size - 1
        while (hi - lo > 1) {
            val mid = (lo + hi) ushr 1
            if (xp[mid] <= v) lo = mid else hi = mid
        }
        val t = (v - xp[lo]) / (xp[hi] - xp[lo])
        return fp[lo] + t * (fp[hi] - fp[lo])
    }
}

/**
 * 迭代基-2 FFT。nFft 是 512（2 的幂），所以够用。
 *
 * 不引第三方 FFT 库：这是唯一需要的信号处理原语，为它加一个依赖
 * 会把「core 模块不依赖任何东西」这条破掉，而那条是它能被 JVM 单测
 * 直接验证的前提。
 */
internal class Fft(private val n: Int) {
    private val levels = Integer.numberOfTrailingZeros(n)
    private val cosT = DoubleArray(n / 2) { cos(2.0 * Math.PI * it / n) }
    private val sinT = DoubleArray(n / 2) { Math.sin(2.0 * Math.PI * it / n) }

    init {
        require(n > 0 && Integer.bitCount(n) == 1) { "FFT 长度必须是 2 的幂，收到 $n" }
    }

    fun transform(re: DoubleArray, im: DoubleArray) {
        // 位反转置换
        for (i in 0 until n) {
            val j = Integer.reverse(i) ushr (32 - levels)
            if (j > i) {
                var t = re[i]; re[i] = re[j]; re[j] = t
                t = im[i]; im[i] = im[j]; im[j] = t
            }
        }
        var size = 2
        while (size <= n) {
            val half = size / 2
            val step = n / size
            var i = 0
            while (i < n) {
                var j = i
                var k = 0
                while (j < i + half) {
                    val l = j + half
                    val tre = re[l] * cosT[k] + im[l] * sinT[k]
                    val tim = -re[l] * sinT[k] + im[l] * cosT[k]
                    re[l] = re[j] - tre; im[l] = im[j] - tim
                    re[j] += tre; im[j] += tim
                    j++; k += step
                }
                i += size
            }
            size *= 2
        }
    }
}
